quimb/evo.py

```diff
@@ -422,6 +436,35 @@ def int_step_callback(t, y, H):
                 step_callback(t, pt, H)
 
         self._step_callback = step_callback
+
+        # if there is an integration stopping condition, set up the function
+        #   to use this
+
+        if int_stop is not None:
+            int_stop_no_tryexcept = int_stop
+
+            # try just passing time and state to a single callback function
+            # if it fails, pass time, state and hamiltonian
+            def int_stop(t, y, H):
+                try:
+                    int_stop_output = int_stop_no_tryexcept(t, y)
+                except TypeError as e:
+                    if 'positional' in e.args[0]:
+                        int_stop_output = int_stop_no_tryexcept(t, y, H)
+                    else:
+                        raise
+                return int_stop_output
+
+            if int_step_callback is not None:
+                int_step_callback_bare = int_step_callback
+
+                def int_step_callback(t, y, H):
```

Comment:
Would be good to call ``pt = qarray(y.reshape(self._d, -1))`` then supply ``pt`` to both ``step_callback`` and ``int_stop``  directly so that the stopping function natively accepts a 'ket' & density matrix as well i.e. not the flattened 1D input.

Reply:
Agreed. Take a look at how I did this within the restructure of the whole function. It now works with density ops (was previously just skipping the test for this)